[Woo Express Cleanup] Delete code related to Free Trial local notification and survey.

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12413
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR removes code related to local notification for Woo Express site creation. One of the notifications provided a survey to the user; the survey is deleted too in this PR.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Since this is mostly deletion, just ensure app still builds fine and there's no test errors, as well as confidence check the deleted code. As part of the Woo Express Cleanup project, site creation is not supported anymore. Since free trial local notifications were created during site creation, the code being deleted here is not reachable by users anyway.

2. To be extra safe, you can also test in a real device with the provisioning profile to test remote notification (e.g: new order notification), to ensure that those notifications still appear. (Check paNNhX-hI-p2 for details) I have tried this and confirmed that remote notifications still appear.

PS: I'll be AFK the whole of next week so if this is needed urgently, please feel free to take over the PR.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 500 lin...
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/196">18.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo commented on WooCommerce/Classes/AppDelegate.swift:
```diff
@@ -69,12 +69,10 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         setupCocoaLumberjack()
         setupLibraryLogger()
         setupLogLevel(.verbose)
-        setupPushNotificationsManagerIfPossible(pushNotesManager, stores: stores)
```
⛔️ We should keep this logic to support remote notification.

itsmeichigo commented on WooCommerce/Classes/AppDelegate.swift:
```diff
@@ -538,7 +508,6 @@ extension AppDelegate {
     /// Runs whenever the Authentication Flow is completed successfully.
     ///
     func authenticatorWasDismissed() {
-        setupPushNotificationsManagerIfPossible(ServiceLocator.pushNotesManager, stores: ServiceLocator.stores)
```
⛔️ We should keep this logic too.

itsmeichigo commented on WooCommerce/Classes/Notifications/LocalNotification.swift:
```diff
@@ -1,180 +0,0 @@
-
-    /// The scenario for the local notification.
-    /// Its raw value is used for the identifier of a local notification and also the event property for analytics.
-    enum Scenario {
```
IMO we should keep this file but remove all the outdated cases related to store creation. The enum will not compile without any case, so we can put a default `unknown` case just for placeholder.

The reason for keeping the file is that we might find ourselves needing local notifications in the future. Keeping the code might help us avoid rewriting the same logic over again while we can avoid doing that in the first place. WDYT?

itsmeichigo commented on WooCommerce/Classes/Notifications/LocalNotificationScheduler.swift:
```diff
@@ -1,57 +0,0 @@
-import Yosemite
-
-/// Handles the scheduling of local notifications with support of remote feature flags.
-final class LocalNotificationScheduler {
```
I'd suggest keeping this file too.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/Settings/SettingsViewController.swift:
```diff
@@ -499,7 +499,6 @@ private extension SettingsViewController {
-            await pushNotesManager.cancelAllNotifications()
             ServiceLocator.stores.deauthenticate()
             navigationController?.popToRootViewController(animated: true)
         }
```
Since we remove the await, we can remove `Task` and move the rest of the code outside the closure to run synchronously.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Upgrades/StorePlanSynchronizer.swift:
```diff
@@ -123,72 +111,6 @@ final class StorePlanSynchronizer: StorePlanSynchronizing {
-}
-
 private extension StorePlanSynchronizer {
     enum Constants {
```
We can remove the enum since it's no longer used.

itsmeichigo reviewed (changes requested):
Thank you Hafiz for working on this cleanup task.

I found a critical issue with the PR - we should not remove the code for registering for remote notifications as that would make remote notifications stop working. You were still able to get notifications when testing because you were likely to install a previous version of the app, where the app already registered for notifications. This PR will not work if you remove the app and reinstall from scratch.

Another comment: I'd suggest keeping the generic logic for local notifications in `LocalNotification`, `LocalNotificationScheduler` and `PushNotificationManager` in case we might need local notifications in the future. Keeping the logic helps us avoid rewriting the logic all over again.

hafizrahman committed:
Delete store creation-related local notifications and free trial survey.

hafizrahman commented:
@itsmeichigo as discussed in Slack, I implemented your suggestions and force-pushed the change.

Some notable items:
- `LocalNotification`, `LocalNotificationScheduler` and `PushNotificationManager` are now preserved
- `setupPushNotificationsManagerIfPossible` is preserved
- the only remaining Scenario is the newly added `unknown(siteID: Int64)` used as example
- other suggestion items are also implemented

I haven't merged the latest `trunk` (just in case another commit reset is needed after further feedback), so will do it after everything's good.

itsmeichigo commented on WooCommerce/Classes/Notifications/LocalNotification.swift:
```diff
@@ -18,51 +19,33 @@ struct LocalNotification {
         }
 
-        /// Helper method to remove postfix from notification identifiers if needed.
+        /// Helper method to remove postfix from nitificatio identifiers if needed.
```
```suggestion
        /// Helper method to remove postfix from notification identifiers if needed.
```

itsmeichigo commented on WooCommerce/Classes/Notifications/LocalNotification.swift:
```diff
@@ -18,51 +19,33 @@ struct LocalNotification {
-                return Identifier.Prefix.freeTrialSurvey24hAfterFreeTrialSubscribed
-            } else if identifier.hasPrefix(Identifier.Prefix.threeDaysAfterStillExploring) {
-                return Identifier.Prefix.threeDaysAfterStillExploring
+            if identifier.hasPrefix(Identifier.Prefix.unknown) {
+                return Identifier.Prefix.unknown
```
Nit: we should not need this prefix check, just returning the identifier should be sufficient.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/12443/commits/ca91d2429c52d062973c89afb2944f84ef95788f

itsmeichigo commented on WooCommerce/Classes/Notifications/LocalNotification.swift:
```diff
@@ -130,51 +101,10 @@ extension LocalNotification {
-                comment: "Message on the local notification to remind after three days."
-            )
-        }
+        static let unknown = NSLocalizedString(
+            "localNotification.Unknown",
+            value: "Unknown",
+            comment: "A placeholder value for local notification"
+        )
```
We should not need to add this for translation, simply setting the title and body to an empty string should be sufficient.

hafizrahman replied:
Handled in https://github.com/woocommerce/woocommerce-ios/pull/12443/commits/ca91d2429c52d062973c89afb2944f84ef95788f

itsmeichigo commented on WooCommerce/WooCommerceTests/Notifications/LocalNotificationTests.swift:
```diff
@@ -3,56 +3,14 @@ import XCTest
 final class LocalNotificationTests: XCTestCase {
 
-    func test_storeCreationComplete_scenario_returns_correct_notification_contents() throws {
+    func test_unknown_scenario_returns_correct_notification_contents() throws {
```
Nit: We should not need tests for the placeholder case.

itsmeichigo replied:
Same comment for the `PushNotificationsManagerTests` file.

hafizrahman replied:
Thanks! Handled in https://github.com/woocommerce/woocommerce-ios/pull/12443/commits/0ee95d1a98a24a83f6e5dcd5c92fe4ba5e09f90e

itsmeichigo reviewed (approved):
Thanks for the updates! I left some nitpicking but pre-approving this PR ✅

hafizrahman committed:
Typo fix

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Update tests to not need handling for placeholder Scenario case (.unknown)

hafizrahman committed:
Various simplification for placeholder case.
